### ISSUE TITLE
Failure to respect module `__getattr__` override

### ISSUE THREAD

**Issue by stefanv** [microsoft/pyright#3989]
We have a [lazy loader](https://github.com/scientific-python/lazy_loader/) used in [Scientific Python](https://scientific-python.org/specs/spec-0001/).

The lazy loader takes `__init__.pyi` stub files, and generates the matching module `__getattr__` and `__all__`.

I would like to tell `pyright` to, when in doubt, respect the imports defined in the `__init__.pyi` to indicate which functions are available in the module. Currently, I suspect it looks at the `__init__.py` and infers that the lazily loaded functions are missing.

### To reproduce

An example package has the following structure:

```
foo
foo/__init__.py
foo/__init__.pyi
foo/bar
foo/bar/__init__.py
foo/bar/__init__.pyi
foo/bar/funcs.py
```

Here is the content of `foo/bar/__init__.pyi`:

```python
from .funcs import somefunc
```

`pyright` complains:

```
error: "somefunc" is not a known member of module (reportGeneralTypeIssues)
```

See [foo.tgz](https://github.com/microsoft/pyright/files/9660185/foo.tgz)
Reproduce by running `pip install lazy-loader` and `pyright`
To see that the package is functioning correctly, run `python test.py`

### Expected behavior

`pyright` sees that a function is defined in a type stub, and assume it is declared *if* there is reasonable doubt on whether it is or not. That doubt could come from, e.g., seeing a module `__getattr__`.

### References

This issue was surfaced by https://github.com/librosa/librosa/pull/1579

**Comment by erictraut** [comment]
[PEP 484](https://peps.python.org/pep-0484/#stub-files), which defines the behavior of type stubs, indicates that a symbol imported by a type stub should be assumed to be private to that stub, not a symbol that is meant to be re-exported by that module. PEP 484 specifies the way to override this default assumption — namely by using a redundant form of an import statement. 

Here's the exact quote:
> Modules and variables imported into the stub are not considered exported from the stub unless the import uses the import ... as ... form or the equivalent from ... import ... as ... form. (UPDATE: To clarify, the intention here is that only names imported using the form X as X will be exported, i.e. the name before and after as must be the same.)

In your code sample, if you change `from .funcs import somefunc` to `from .funcs import somefunc as somefunc`, pyright (and other type checkers) will then assume that `somefunc` is intended to be exported by `__init__.pyi`.

So this isn't a bug. Pyright is working according to the typing standard specifications.

**Comment by stefanv** [comment]
Thanks @erictraut, that was very helpful!
